### PR TITLE
Catch errors when JSON is invalid

### PR DESCRIPTION
**Pull request by metsma**
WE2-925

Signed-off-by: Raul Metsma <raul@metsma.ee>

### CODE REVIEW

File: CustomAction.config
Comment (mrts) [inline CustomAction.config:1]: Why did you remove `src`?

File: FirefoxAction.cs [hunk 1]
```diff
@@ -36,44 +40,67 @@ public static ActionResult ExtensionSettingsInstall(Session session)
...
+                    firefox.SetValue(KEY_NAME, json.ToString().Split('\n'));
                 }
-                return ActionResult.Success;
             }
+            return ActionResult.Success;
         }
 
-        private static (string UUID, string URL) GetExtensionSettingsFromSession(Session session)
+        [CustomAction]
+        public static ActionResult ExtensionSettingsFix(Session session)
```
Comment (mrts) [inline FirefoxAction.cs]: We need to document why this is required as it looks quite scary - we are deleting a registry key that we do not own, something in the lines of the following:

_For unknown reasons, the Firefox ExtensionSettings registry entry sometimes contains corrupt JSON data, which causes errors during installer operations. In this case, this action removes the entry._

README should be updated as well.

File: FirefoxAction.cs [hunk 2]
```diff
@@ -101,7 +125,7 @@ internal static RegistryKey OpenOrCreateSubKey(this RegistryKey registryKey, str
 
         internal static string GetStringValue(this RegistryKey registryKey, string name, string defaultValue = null)
         {
-            if (!registryKey.ContainsName(name))
+            if (Array.IndexOf(registryKey.GetValueNames(), name) < 0)
```
Comment (mrts) [inline FirefoxAction.cs]: Would LINQ be more idiomatic here, also, let's use braces throughout:

```suggestion
            if (registryKey.GetValueNames().Any(name.Equals)) {
```

File: FirefoxAction.cs [hunk 3]
```diff
@@ -114,14 +138,21 @@ internal static string GetStringValue(this RegistryKey registryKey, string name,
             }
         }
 
-        internal static bool ContainsName(this RegistryKey key, string name)
+        internal static JObject GetJSON(this RegistryKey registryKey, string name, string defaultValue = null)
```
Comment (mrts) [inline FirefoxAction.cs:119]: This is quite neat. It is not easy to follow, but the `new JObject()` return works subtly in both `ExtensionSettingsRemove` and `ExtensionSettingsInstall` cases, which is cool.

File: CustomAction.config
Comment (metsma) [reply CustomAction.config:1]: It is alien in C# project and same level folder gives access to README.md in project and licence file

**Review by mrts — APPROVED** [review]
